### PR TITLE
Build with Scala-3

### PR DESCRIPTION
**Pull request by hamnis**
Upgrade libraries
Move shapeless code into scala-2 source directory
Use scala.derving.Mirror for scala-3 code

### CODE REVIEW

File: src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/SnowplowEvent.scala [hunk 1]
```diff
@@ -71,26 +74,33 @@ object SnowplowEvent {
...
-    Encoder.instance { contexts: Contexts =>
-      if (contexts.data.isEmpty) JsonObject.empty.asJson
-      else
-        JsonObject(
-          ("schema", Common.ContextsUri.toSchemaUri.asJson),
-          ("data", contexts.data.asJson)
-        ).asJson
-    }
+  object Contexts {
+    implicit final val contextsCirceEncoder: Encoder[Contexts] =
```
Comment (hamnis) [inline src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/SnowplowEvent.scala:78]: to help with implicit resolving.

File: src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/SnowplowEvent.scala [hunk 2]
```diff
@@ -30,34 +29,38 @@ object SnowplowEvent {
...
-  implicit final val unstructCirceEncoder: Encoder[UnstructEvent] =
-    Encoder.instance { unstructEvent: UnstructEvent =>
-      if (unstructEvent.data.isEmpty) Json.Null
-      else
-        JsonObject(
-          ("schema", Common.UnstructEventUri.toSchemaUri.asJson),
-          ("data", unstructEvent.data.asJson)
-        ).asJson
-    }
+  object UnstructEvent {
```
Comment (hamnis) [inline src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/SnowplowEvent.scala:42]: to help with implicit resolving.

File: .scalafmt.conf [hunk 3]
```diff
@@ -1,23 +1,30 @@
-version = "2.6.1"
+version = "3.5.9"
```
Comment (hamnis) [inline .scalafmt.conf:1]: this including all formatting should probably be its own PR.

File: src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/SnowplowEvent.scala [hunk 4]
```diff
@@ -71,26 +74,33 @@ object SnowplowEvent {
...
+        else
+          JsonObject(
+            ("schema", Common.ContextsUri.toSchemaUri.asJson),
+            ("data", contexts.data.asJson)
+          ).asJson
+      }
 
-  implicit val contextsDecoder: Decoder[Contexts] = deriveDecoder[Contexts].recover {
-    case DecodingFailure(_, DownField("data") :: _) => Contexts(List())
+    implicit val contextsDecoder: Decoder[Contexts] = Decoder.forProduct1("data")(apply).recover {
```
Comment (hamnis) [inline src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/SnowplowEvent.scala:88]: deriveEncoder is really overkill here, and potentially scary.

**Comment by istreeter** [comment]
Hi @hamnis, this looks great!  Thank you for yet another excellent contribution.

My only concern is about moving decoders/encoders into the companion objects.  While I know this is a better way to organise code, and better for resolving implicits, it unfortunately breaks binary compatibility with the previous release of this package.

Please can you check #126 where I amended your commit to put the encoders/decoders back to their original place?  If you are happy with #126 then I will get it merged and released.  We were planning a 3.1.0 version soon anyway.

**Comment by hamnis** [comment]
LGTM, closing this in favour of #126
